Remove 50ms delay at exit for programs using systhreads

v-gb opened the pull request:
The following program is surprisingly slow:

```
$ cat /tmp/b.ml
let () = Thread.join (Thread.create ignore ())
let () = for _ = 0 to 100000; do () done
$ ocamlopt -I +threads unix.cmxa threads.cmxa /tmp/b.ml -o b
$ time ./b

real	0m0.053s
user	0m0.000s
sys	0m0.000s
```

The reason is that the systhreads library registers an at_exit handler that waits for the tick thread to stop, but doesn't interrupt its select syscall. In the example above, the tick thread just started the select, and so we wait the full 50ms.

This is annoying for short lived async program, because many things cause a thread to be created, like writing to a terminal. The program that started my investigation spent 100ms doing real work and 30ms waiting for the tick thread to stop, ie 25% of its time.

There are two ways of fixing this: either interrupt the thread (self-pipe trick or signal), or don't stop it. The former apparently needs a different implementation for unix and windows, so I went with the latter. I made a variation of [at_exit] that tells the callback whether the process is exiting or only the ocaml runtime is exiting, and I use it in the thread library to skip the waiting when the process is exiting.

With this, the program above takes 3ms.

v-gb committed:
Remove 50ms delay at exit for programs using threads

$ cat /tmp/b.ml
let () = Thread.join (Thread.create ignore ())
let () = for _ = 0 to 100000; do () done
$ ocamlopt -I +threads unix.cmxa threads.cmxa /tmp/b.ml -o b
$ time ./b # before this commit

real	0m0.053s
user	0m0.000s
sys	0m0.000s
$ time ./b # after this commit

real	0m0.003s
user	0m0.000s
sys	0m0.000s

mshinwell commented:
Contentious point first: does anyone object to changing the interface of `Pervasives` in this way?

v-gb commented:
I'll note a couple of things:
- I changed the documentation of `at_exit`, which was wrong since the introduction of `caml_shutdown`
- I don't know why a comment in threads I changed talks about DLL unloading, given that the dynlink module doesn't offer this functionality

stedolan commented:
I don't think it's a good idea to complicate Pervasives with two different types of termination, especially as the real fix (getting the tick thread to terminate promptly) isn't all that difficult.

> There are two ways of fixing this: either interrupt the thread (self-pipe trick or signal), or don't stop it.

There is a simpler way. Instead of using `select()` to implement the timer delay, we could make a condition variable and a terminated flag, and use `pthread_cond_timedwait`. To shutdown, set the flag, signal the condition variable and join (don't kill) the ticker thread.

> The former apparently needs a different implementation for unix and windows, so I went with the latter.

All of the systhreads timer-ticking logic is implemented separately for unix and windows, but this isn't too much of a hardship. `pthread_cond_timedwait` doesn't exist on Windows, but `WaitForSingleObject` can be used to the same effect (in this case).

v-gb commented:
Thanks for taking a look.
I disagree that stopping the tick thread is "the real fix". No one is joining all user thread on exit or deallocating the runtime on exit. I don't see why the tick thread should be so different.

If the change to pervasives is what you don't like, then I don't mind removing it, removing the at_exit handler from the threads library, and instead having some kind of hook in C that caml_shutdown will call to stop the tick thread.
This way the tick thread will be treated the same as all the other resources of the runtime.

(I also find stopping the tick thread slightly dubious given that some people run ocaml code after calling Pervasives.do_at_exit)

v-gb commented:
I made the pull request not change pervasives, but the implementation strategy is the same: don't stop the tick thread on exit.

v-gb commented:
@stedolan @mshinwell any opinion on the new version?

gasche commented:
@stedolan I'm willing to merge or not following your decision if you have time to handle this PR.

mshinwell commented:
I've asked @stedolan if he can handle the review of this one.

stedolan commented:
Apologies for the delay! This version looks good to me. @gasche, please merge!

I'm still midly uneasy about having separate callbacks for `caml_shutdown` and `exit` (my understanding was that the goal of `caml_shutdown` was to be semantically `exit` without terminating the process), but in this version they are clearly an internal implementation detail rather than a public API.

> removing the at_exit handler from the threads library, and instead having some kind of hook in C that caml_shutdown will call to stop the tick thread.

The systhreads library interacts with the runtime using a mixture of C hooks and primitives called by `thread.ml`. Since the tick thread is started by a call to a primitive from `thread.ml`, it's cleaner to have it terminated by the same mechanism. So, I think the current mechanism of triggering cleanup via thread.ml is better than using a hook in C.

> (I also find stopping the tick thread slightly dubious given that some people run ocaml code after calling Pervasives.do_at_exit)

I think the dubious thing is calling Pervasives.do_at_exit, which is clearly marked as "for system use only"!

gasche reviewed (approved):
Approved based on @stedolan's review.

gasche commented:
@sliquister would you add Stephen Dolan as reviewer in the Changes entry? Then I'll merge.

v-gb commented:
@stedolan Thanks. 
@gasche done.
